cmd/gpu-kubelet-plugin/driver.go

```diff
@@ -177,6 +202,68 @@ func (d *driver) nodeUnprepareResource(ctx context.Context, claimNs kubeletplugi
 	return nil
 }
 
+func (d *driver) deviceHealthEvents(ctx context.Context, nodeName string) {
+	klog.Info("Processing device health notifications")
+	for {
+		select {
+		case <-ctx.Done():
+			klog.V(6).Info("Stop processing device health notifications")
+			return
+		case device, ok := <-d.deviceHealthMonitor.Unhealthy():
+			if !ok {
+				klog.V(6).Info("Health monitor channel closed")
+				return
+			}
+			uuid := device.UUID()
+
+			klog.Warningf("Received unhealthy notification for device: %s", uuid)
+
+			if !device.IsHealthy() {
+				klog.V(6).Infof("Device: %s is aleady marked unhealthy. Skip republishing resourceslice", uuid)
+				continue
+			}
+
+			release, err := d.pulock.Acquire(ctx, flock.WithTimeout(10*time.Second))
+			if err != nil {
+				klog.Errorf("error acquiring prep/unprep lock for health status update: %v", err)
+				continue
+			}
+
+			// Mark device as unhealthy.
+			d.state.UpdateDeviceHealthStatus(device, Unhealthy)
+
+			// Republish resource slice with only healthy devices
+			// There is no remediation loop right now meaning if the unhealthy device is fixed,
+			// driver needs to be restarted to publish the resourceslice with all devices
+			var resourceSlice resourceslice.Slice
+			for _, dev := range d.state.allocatable {
+				uuid := dev.UUID()
+				if dev.IsHealthy() {
+					klog.V(6).Infof("Device: %s is healthy, added to resoureslice", uuid)
+					resourceSlice.Devices = append(resourceSlice.Devices, dev.GetDevice())
+				} else {
+					klog.Warningf("Device: %s is unhealthy, will be removed from resoureslice", uuid)
+				}
+			}
+
+			klog.V(6).Info("Rebulishing resourceslice with healthy devices")
+			resources := resourceslice.DriverResources{
+				Pools: map[string]resourceslice.Pool{
+					nodeName: {Slices: []resourceslice.Slice{resourceSlice}},
+				},
+			}
+
+			if err := d.pluginhelper.PublishResources(ctx, resources); err != nil {
+				klog.Errorf("Failed to publish resources after device health status update: %v", err)
+			} else {
+				klog.V(6).Info("Successfully republished resources without unhealthy device")
+			}
+
+			release()
```

Comment:
If we factor this logic into a function, then we could use a deferred call to `release()` after taking the lock. This may be less error prone if we do ever add code paths that return from this logic.